scrypto-test/src/environment/client_api.rs

```diff
@@ -44,16 +44,56 @@ macro_rules! implement_client_api {
         ),* $(,)*
     ) => {
         $(
-            impl $trait<RuntimeError> for TestEnvironment {
+            impl<D> $trait<RuntimeError> for TestEnvironment<D>
+            where
+                D: SubstateDatabase + CommittableSubstateDatabase + 'static
+            {
                 $(
                     #[inline]
                     fn $func_ident(&mut self, $($input_ident: $input_type),*) -> $outputs {
-                        self.0.with_kernel_mut(|kernel| {
+                        let logs_before = self.0.with_kernel_mut(|kernel| {
+                            kernel
+                                .kernel_get_system_state()
+                                .system
+                                .modules
+                                .transaction_runtime()
+                                .map(|runtime| runtime.logs.len())
+                                .unwrap_or(0)
+                        });
+
+                        let rtn = self.0.with_kernel_mut(|kernel| {
                             SystemService {
                                 api: kernel,
                                 phantom: PhantomData,
                             }.$func_ident( $($input_ident),* )
-                        })
+                        });
+
+                        self.0.with_kernel_mut(|kernel| {
+                            let logs_after = kernel
+                                .kernel_get_system_state()
+                                .system
+                                .modules
+                                .transaction_runtime()
+                                .map(|runtime| runtime.logs.len())
+                                .unwrap_or(0);
+
+                            if logs_before != logs_after {
+                                for (level, message) in kernel
+                                    .kernel_get_system_state()
+                                    .system
+                                    .modules
+                                    .transaction_runtime()
+                                    .map(|module| module.logs.clone())
```

Comment:
Minor: might be able to avoid cloning all the logs?

Reply:
Yup, I've removed this clone